File: competitive-programming/src/edu/algos/li20_dp/medium/Grid_MultiStartNEnd_L931_MinimumFallingPathSum.java
package edu.algos.li20_dp.medium;

public class Grid_MultiStartNEnd_L931_MinimumFallingPathSum {
    public static void main(String[] args) {
        Grid_MultiStartNEnd_L931_MinimumFallingPathSum obj = new Grid_MultiStartNEnd_L931_MinimumFallingPathSum();

        int[][] matrix = new int[][]{{2,1,3},{6,5,4},{7,8,9}};
        System.out.println(obj.minFallingPathSum(matrix));
    }

    // Problem: https://leetcode.com/problems/minimum-falling-path-sum/
    // Idea: Same as triangle problem, but there can be multiple source. Hence, we need to start recursion from each column and take max of all the col maxes.
    public int minFallingPathSum(int[][] matrix) {
        int m=matrix.length, n=matrix[0].length;

        int[][] cost = new int[m][n];
        for(int j=0; j<n; j++){
            cost[0][j] = matrix[0][j];
        }
        for(int i=1; i<m; i++){
            for(int j=0; j<n; j++){
                int l = matrix[i][j] + cost[i-1][j];
                int ld = (j-1>=0)?  matrix[i][j] + cost[i-1][j-1] : (int)1e9;
                int rd = (j+1<m)?  matrix[i][j] + cost[i-1][j+1] : (int)1e9;

                cost[i][j] = Math.min(l, Math.min(ld, rd));
            }
        }

        int min = (int)1e9;
        for(int j=0; j<n; j++){
            min = Math.min(min, cost[n-1][j]);
        }
        return min;
    }

    public int minFallingPathSumRec(int[][] matrix) {
        int min = (int)1e8, m=matrix.length;
        for(int j=0; j<matrix[0].length; j++){
            min = Math.min(min, minFallingPathSum(matrix, m-1, j));
        }
        return min;
    }

    private int minFallingPathSum(int[][] matrix, int i, int j) {
        if(j<0 || j>=matrix[0].length)
            return (int)1e8;
        if(i==0)
            return matrix[i][j];

        int l = matrix[i][j] + minFallingPathSum(matrix, i-1, j);
        int ld = matrix[i][j] + minFallingPathSum(matrix, i-1, j-1);
        int rd = matrix[i][j] + minFallingPathSum(matrix, i-1, j+1);

        return Math.min(l, Math.min(ld, rd));
    }
}

File: dsa/topics/15_dp/python/grid_multi_start_nend_l931_minimum_falling_path_sum.py
# Problem: https://leetcode.com/problems/minimum-falling-path-sum/
# Idea: Same as triangle problem, but there can be multiple source. Hence, we need to start recursion from each column and take max of all the col maxes.
# Time: O(m*n)
# Space: O(m*n)
def minFallingPathSum(matrix):
    m, n = len(matrix), len(matrix[0])

    cost = [[0] * n for _ in range(m)]

    for j in range(n):
        cost[0][j] = matrix[0][j]  # i=0

    for i in range(1, m):
        for j in range(n):
            l = matrix[i][j] + cost[i - 1][j]
            ld = matrix[i][j] + cost[i - 1][j - 1] if j - 1 >= 0 else int(1e9)
            rd = matrix[i][j] + cost[i - 1][j + 1] if j + 1 < n else int(1e9)

            cost[i][j] = min(l, ld, rd)

    ans = int(1e9)
    for j in range(n):
        ans = min(ans, cost[m - 1][j])

    return ans


def minFallingPathSum2(matrix):
    m = len(matrix)
    ans = int(1e8)

    for j in range(len(matrix[0])):
        ans = min(ans, minFallingPathSum_recursive(matrix, m - 1, j))

    return ans


def minFallingPathSum_recursive(matrix, i, j):
    if j < 0 or j >= len(matrix[0]):
        return int(1e8)
    if i == 0:
        return matrix[i][j]

    l = matrix[i][j] + minFallingPathSum_recursive(matrix, i - 1, j)
    ld = matrix[i][j] + minFallingPathSum_recursive(matrix, i - 1, j - 1)
    rd = matrix[i][j] + minFallingPathSum_recursive(matrix, i - 1, j + 1)

    return min(l, ld, rd)


# ---------------- Test (same as Java main) ----------------

matrix = [
    [2, 1, 3],
    [6, 5, 4],
    [7, 8, 9]
]

print(minFallingPathSum(matrix))
print(minFallingPathSum2(matrix))
